### PR TITLE
Feat(Setting) donor management module

### LINKED ISSUE #4826: Specify the donor on Stock entry from donation
We currently don't define the donator when we do the stock entry by donation. the donator details may also be useful at some point.

### PR DESCRIPTION
**Pull request by jeremielodi**
closes #4826

### CODE REVIEW

**Comment by jeremielodi** [comment]
There are still others step on which my working in order to totally fix #4826

**Comment by jmcameron** [comment]
I did a functional test with Bhima_test and found that the donor creation/editing works fine.  Thanks.   I have not scanned the code carefully.

I assume that future Pull Requests will deal with these issues:

- Adding more details to the Donor (eg, address, Department, Company, contact info, etc).
- Tying this into the Stock Entry module so that you can select an existing Donor or create a new one without leaving the Stock Entry form

**Comment by jmcameron** [comment]
I did another functional test (I did not do a detailed code review).   I was able to create, edit, and delete donors.  I think that the additional donor contact details will be useful.  It will be hard to tell what is really needed for details until we use this in the field with actual donors, etc.

However, I did find one problem:  I was able to create a Donor, but when I went to Stock > Stock Entry, and selected "DONOR", I did not see my new donor in the list.

**Comment by jeremielodi** [comment]
The only thing that remains is to create a donation from find modal. I will work on that today and ask for a review just after

File: client/src/modules/donor/donation.service.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+angular.module('bhima.services')
+  .service('DonationService', DonationService);
+
+DonationService.$inject = ['PrototypeApiService'];
+
+/**
+ * Role Service
+ *
+ * A service wrapper for the /donors HTTP endpoint.
```
Comment (mbayopanda) [inline client/src/modules/donor/donation.service.js:9]: This service is for `/donors` or `/donations` ? If it is for `/donations` you could update this line

File: client/src/modules/donor/donor.ctrl.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+DonorController.$inject = [
+  '$uibModal', 'DonorService', 'ModalService',
+  'NotifyService', 'uiGridConstants',
+];
+
+function DonorController($uibModal, Donor, Modal, Notify, uiGridConstants) {
+  const vm = this;
+
+  vm.createUpdateDonorModal = (selectedRole = {}) => {
```
Comment (mbayopanda) [inline client/src/modules/donor/donor.ctrl.js:12]: I think this parameter is not related to role may be `selectedDonor` is a better name

File: client/src/modules/donor/donor.ctrl.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+          .catch(Notify.handleError);
+      });
+  };
+
+  function load() {
+    vm.loading = true;
+
+    Donor.read()
+      .then(roles => {
+        vm.gridOptions.data = roles;
```
Comment (mbayopanda) [inline client/src/modules/donor/donor.ctrl.js:41]: `donors` in place of `roles` will be better

File: client/src/modules/donor/modal/createUpdate.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
+angular.module('bhima.controllers')
+  .controller('DonorAddController', DonorAddController);
```
Comment (mbayopanda) [inline client/src/modules/donor/modal/createUpdate.js:2]: It seems like the controller is not only for creation of donors, so you could rename this controller to something like : `DonorCreateAndUpdateController`, `DonorManagementController`, etc.

File: client/src/modules/stock/entry/entry.js [hunk 5]
```diff
@@ -547,14 +557,7 @@ function StockEntryController(
...
 
-    /*
-      the origin_uuid of lots is set on the client
-      because donation table depends on donor, and donor management
-      is not yet implemented in the application
-
-      TODO: add a donor management module
-    */
-    movement.lots = Stock.processLotsFromStore(vm.stockForm.store.data, Uuid());
+    movement.lots = Stock.processLotsFromStore(vm.stockForm.store.data, vm.movement.entity.uuid);
```
Comment (mbayopanda) [inline client/src/modules/stock/entry/entry.js:560]: 👍

File: client/src/modules/stock/entry/modals/findDonation.modal.html [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        </button>
+      </div>
+    </div>
+  </div>
+
+  <div class="modal-body">
+    <div class="row">
+      <div class="col-xs-12">
+        <div
+          id="PurchaseGrid"
```
Comment (mbayopanda) [inline client/src/modules/stock/entry/modals/findDonation.modal.html]: The grid is not for purchase I think, maybe `DonationGrid` will be better

File: client/src/modules/stock/entry/modals/findDonation.modal.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  vm.gridOptions = { appScopeProvider : vm };
+
+  const filtering = new Filtering(vm.gridOptions);
+
+  const columns = [
+    {
+      field            : 'reference',
+      displayName      : 'TABLE.COLUMNS.REFERENCE',
+      headerCellFilter : 'translate',
+      cellTemplate     : 'modules/stock/entry/modals/templates/purchase_reference.tmpl.html',
```
Comment (mbayopanda) [inline client/src/modules/stock/entry/modals/findDonation.modal.js]: It seems like this template is used also with purchase, do you think we can rename it ?

File: client/src/modules/stock/entry/modals/findDonation.modal.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  /** toggle filter */
+  function toggleFilter() {
+    vm.filterEnabled = !vm.filterEnabled;
+    vm.gridOptions.enableFiltering = vm.filterEnabled;
+    vm.gridApi.core.notifyDataChange(uiGridConstants.dataChange.COLUMN);
+  }
+
+  /** get purchase document */
```
Comment (mbayopanda) [inline client/src/modules/stock/entry/modals/findDonation.modal.js]: Is it possible to add comments about the use of `Receipts.purchase` here.

Also I cannot get the receipt of donation

File: server/config/routes.js [hunk 9]
```diff
@@ -564,6 +566,20 @@ exports.configure = function configure(app) {
...
+  app.put('/donors/:id', donor.update);
+  app.delete('/donors/:id', donor.remove);
+
+  // donor controller
+  app.get('/donations', donation.read);
+  app.post('/donations', donation.create);
+  app.get('/donations/:uuid', donation.detail);
+  app.put('/donations/:uuid', donation.update);
+  app.delete('/donations/:uuid', donation.remove);
+
```
Comment (mbayopanda) [inline server/config/routes.js:584]: 👍

File: client/src/modules/donor/modal/createUpdate.html [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+        <label class="label label-warning text-uppercase" translate>
+          {{DonorAddCtrl.action}}
+        </label>
+      </li>
+    </ol>
+  </div>
+
+  <div class="modal-body">
+    <div class="row">
```
Comment (mbayopanda) [inline client/src/modules/donor/modal/createUpdate.html:15]: Could you remove this `<div class="row">` it is not necessary

**Review by mbayopanda — COMMENTED** [review]
👍 Good job @jeremielodi most part of the work that you made work fine

I have just notice that : 
- We cannot get the donation receipt from the donation modal
- There is not yet a donation module for creating donation
- We cannot populate the grid with donation data in the stock entry module

**Review by jmcameron — APPROVED** [review]
I did another test and the donor add/edit/delete works fine.
However, using the Donors in the Stock Entry page is confusing to me.

- When I added a new donor, and did Stock > Stock Entry, clicked on [Donor], the new Donor is not listed.
- In the pop-up dialog that comes up to select a donor, if I click the check mark for the existing Donor and click on [Submit], I return to the Stock Entry page, but no Donor is indicated in the [Donor] button AND there is still a warning at the bottom "No source selected".
- In the pop-up donor selection dialog, there is an [+] button on the top right to add a Donor (right?).  However, I think there is a problem with the controls:
image
Notice that controls to add stock items appears.  That should not be in this form. This form should just select the donor and return the the main Stock Entry page with the Donor inserted.   If you click on the [Add] button, to add stock items (which you should not do in this form), you will see error messages on the underlying screen, not this dialog!  So I think that the lower part (to add stock items) should be removed from this form.
- Finally, is the pop-up Donor selection dialog the only place to associated a donation with a project?

What would make more sense to me (and probably be simpler for users), would be for 

- The Donor selection dialog allows you to pick the donor and return to the main Stock entry form.
- The Donor would be selected and displayed in the [Donor] button.
- Two new fields would appear in the Stock Entry form: 
    - a task selection  field (drop-down menu) 
    - a donation description text field
- No additional donation date field is needed (probably); I think that the date of the entry should be adequate.  One could enter any other original donation dates, etc, in the Donation description field.

**Comment by jeremielodi** [comment]
The field ```date``` is enough for specifying the donation date, the ```stock_movement``` table has another column ```created_at```  that tell us the date the stock has been encoded

**Comment by jeremielodi** [comment]
The donor list shows up without error
donors

**Comment by jmcameron** [comment]
> 
> 
> The donor list shows up without error

Yes, I saw that.  And it worked for me, but I thought the process was difficult.   I outlined the way that seemed more user-friendly to me in my previous post.

**Comment by jeremielodi** [comment]
Even for purshase order if you create a supplier no source will be selected in the stock entry, @mbayopanda can tell us more about this

**Comment by mbayopanda** [comment]
> Even for purshase order if you create a supplier no source will be selected in the stock entry, @mbayopanda can tell us more about this

I agree that we can simplify this feature by having just the donor (when we click on the donation button), we have already these pieces of information (date, description, and inventories) in the stock entry form, and I am not sure if we really need the project here.

File: test/end-to-end/stock/stock.z2.inventory-adjustment-inventories.spec.js [hunk 11]
```diff
@@ -25,7 +25,7 @@ function StockInventoriesRegistryTests() {
   it('find 3 inventory in Depot Principal plus one line for the Grouping', async () => {
     await modal.setDepot('Depot Principal');
     await modal.submit();
-    await GU.expectRowCount(gridId, 2 + GROUPING_ROW);
+    await GU.expectRowCount(gridId, 3 + GROUPING_ROW);
```
Comment (jniles) [inline test/end-to-end/stock/stock.z2.inventory-adjustment-inventories.spec.js:28]: image

Just FYI, when I run this on my machine, this is the state of the registry when we get to this failing test.

**Comment by jniles** [comment]
@jeremielodi do you think it would be possible to get this in by the end of this week?

**Comment by jniles** [comment]
@jeremielodi what is the status of this PR?
